### PR TITLE
Spotify Search Functionality

### PR DESCRIPTION
**Pull request by Ben2356**
This functionality utilizes the Spotify Web API (user must be authorized) to be able to search for Tracks, Artists, Albums, Podcasts, and Shows on Spotify. Once the user has entered more than 2 characters the current query string will be searched and the search suggestions will be updated in the suggestion model as they are fetched. The search suggestion screen isn't very flexible with how it displays results so when the user clicks OK on the input state they will be taken to BrowseView that shows the search results with cover art and formatted similarly to how the Spotify app displays results.

Demo of Car UI: 
https://user-images.githubusercontent.com/11298573/106551348-a6dbc180-64da-11eb-8774-99d7d958307b.mp4

I've also added a "Search" tab to the phone UI that allows for simple searching of a query.
Demo of PhoneUI:
https://user-images.githubusercontent.com/11298573/106551340-a0e5e080-64da-11eb-9483-cade5b0c922c.mp4

### CODE REVIEW

File: app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt [hunk 1]
```diff
@@ -422,7 +434,7 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
 			MusicAction.SKIP_TO_QUEUE_ITEM -> false
 			MusicAction.SET_SHUFFLE_MODE -> playerActions?.canToggleShuffle == true
 			MusicAction.SET_REPEAT_MODE -> playerActions?.canRepeatContext == true || playerActions?.canRepeatTrack == true
-			// figure out search
+			MusicAction.PLAY_FROM_SEARCH -> true
```
Comment (hufman) [inline app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:437]: The SpotifyAppController shouldn't declare support for this because its `playFromSearch` function doesn't do anything. This menu entry is shown because the GenericAppController controlling Spotify's MediaSession provides the functionality for `playFromSearch`.

File: app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt [hunk 2]
```diff
@@ -505,15 +517,34 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
...
-		// requires a Web API call, not sure how to get the access token
-		return null
+	/**
+	 * Removes the shuffle play button [MusicMetadata] if it is present in the supplied list.
+	 *
+	 * When loading a list of tracks such as an album, a shuffle play button [MusicMetadata] object is
+	 * sometimes present. Call this method to get the list of [MusicMetadata]s that omits the shuffle
+	 * play button.
+	 */
+	private fun removeShufflePlayButtonMetadata(items: List<MusicMetadata>): List<MusicMetadata> {
```
Comment (hufman) [inline app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:534]: This tweak should be pulled out to a separate PR, because it doesn't really relate to the Search functionality.

File: app/src/main/java/me/hufman/androidautoidrive/music/spotify/SpotifyAuthStateManager.kt [hunk 3]
```diff
@@ -96,12 +97,20 @@ class SpotifyAuthStateManager private constructor(val appSettings: MutableAppSet
...
 
 	/**
-	 * Updates the [AuthState] with the token response and authorization exception, replacing the old
-	 * AuthState in the shared preferences file with the new one.
+	 * Updates the [AuthState] with the provided [Token], replacing the old AuthState with the new one.
 	 */
-	fun updateTokenResponse(response: TokenResponse?, ex: AuthorizationException?) {
-		currentState.update(response, ex)
-		writeState(currentState)
+	fun updateTokenResponseWithToken(token: Token, clientId: String) {
```
Comment (hufman) [inline app/src/main/java/me/hufman/androidautoidrive/music/spotify/SpotifyAuthStateManager.kt:102]: Ooo does this seem to help the spontaneous logouts? Feel free to pull this out to a separate PR too

File: app/src/main/res/layout/music_search_container.xml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (hufman) [inline app/src/main/res/layout/music_search_container.xml:1]: I think this file isn't used

File: app/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -257,4 +261,6 @@
     <string name="MUSIC_SPOTIFY_THUMBS_UP_SELECTED">Thumbed Up</string>
     <string name="MUSIC_SPOTIFY_THUMB_DOWN">Thumb Down</string>
     <string name="MUSIC_SPOTIFY_THUMBS_DOWN_SELECTED">Thumbed Down</string>
+    <!-- TODO: Remove or change this placeholder text -->
+    <string name="hello_blank_fragment">Hello blank fragment</string>
```
Comment (hufman) [inline app/src/main/res/values/strings.xml:265]: I think this TODO was missed :)

File: app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/BrowsePageView.kt [hunk 6]
```diff
@@ -373,23 +399,30 @@ class BrowsePageView(val state: RHMIState, val musicImageIDs: MusicImageIDs, val
 				if (item == SEARCHRESULT_EMPTY || item == SEARCHRESULT_SEARCHING) {
 					// invalid selection, don't change states
 					inputComponent.getSuggestAction()?.asHMIAction()?.getTargetModel()?.asRaIntModel()?.value = 0
+
 					throw RHMIActionAbort()
 				} else if (item == BrowseView.SEARCHRESULT_PLAY_FROM_SEARCH) {
 					browseController.playFromSearch(this.input)
 					browseController.onListSelection(item, inputComponent.getSuggestAction()?.asHMIAction())
 				} else {
```
Comment (hufman) [inline app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/BrowsePageView.kt:407]: Can a new SEARCHRESULT be added to the list to say something like "View Full Results" which would do the same `showSearchResults` as the Ok button?

File: app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/BrowsePageView.kt [hunk 7]
```diff
@@ -373,23 +399,30 @@ class BrowsePageView(val state: RHMIState, val musicImageIDs: MusicImageIDs, val
...
 				return if (row.subtitle != null) {
-					UnicodeCleaner.clean("${row.title}\n${row.subtitle}")
+					UnicodeCleaner.clean("${row.title} - ${row.subtitle} - ${row.artist}")
 				} else {
 					UnicodeCleaner.clean(row.title ?: "")
 				}
 			}
+
+			override fun onOk() {
+				browseController.showSearchResults(deferredSearchResults, inputComponent.getResultAction()?.asHMIAction())
```
Comment (hufman) [inline app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/BrowsePageView.kt:424]: Perhaps this could be conditional? I could imagine the scenario where the deferredSearchResults are complete but empty, and I could imagine the desired `onOk` behavior in this situation would be to `PlayFromSearch` if available.

File: app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/BrowsePageView.kt [hunk 8]
```diff
@@ -28,7 +28,7 @@ enum class BrowseAction(val getLabel: () -> String) {
 		return getLabel()
 	}
 }
-class BrowsePageView(val state: RHMIState, val musicImageIDs: MusicImageIDs, val browsePageModel: BrowsePageModel, val browseController: BrowsePageController, var previouslySelected: MusicMetadata?, val graphicsHelpers: GraphicsHelpers): CoroutineScope {
+class BrowsePageView(val state: RHMIState, musicImageIDs: MusicImageIDs, val browsePageModel: BrowsePageModel, val browseController: BrowsePageController, var previouslySelected: MusicMetadata?, val graphicsHelpers: GraphicsHelpers): CoroutineScope {
```
Comment (hufman) [inline app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/BrowsePageView.kt:31]: I'm preparing a refactor of the BrowsePageView on the [browsepage_model](https://github.com/hufman/AndroidAutoIdrive/compare/browsepage_model) branch, which simplifies this class a lot and remove any special-casing for this extra Search Results usecase. After I finish that, this module should be much more straight-forward.

**Comment by hufman** [comment]
Can you also change the MusicPermissionsFragment layout to show a different string like `Ability to search Spotify` to emphasize the user benefit of clicking the button?

File: app/src/main/res/values/strings.xml (hunk 5 above)
Comment (Ben2356) [reply app/src/main/res/values/strings.xml]: Yep that was, removed that :)

File: app/src/main/res/layout/music_search_container.xml (hunk 4 above)
Comment (Ben2356) [reply app/src/main/res/layout/music_search_container.xml]: Yeah you're right, I must have missed removing this file. Removed it

**Comment by hufman** [comment]
The SpotifyAppController.Connector has logic to trigger MusicAppDiscovery again after logging in with the control API to add the Browseable flag. Can you add similar logic to add the Searchable flag after the Web API is authenticated?

File: app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt [hunk 9]
```diff
@@ -505,15 +517,34 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
...
+		return if (items.isNotEmpty() && items[0].artist == "" && items[0].title == L.MUSIC_SHUFFLE_PLAY_BUTTON_TITLE) {
```
Comment (hufman) [inline app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:535]: I like this solution, comparing the menu entry to a translated string to know when to omit it. However, the string comes from Spotify and follows the phone's translations, while the `L` object might not (defaults to the car's language, and can be manually set to any supported language).
Instead of adding to the `L` object, this should use `context.getString()` to load the phone language's translation of the string (perhaps at object initialization time). The string id should then use snake_case to match the other phone strings.

Comment (hufman) [reply app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:535]: Alternativelly, is there some clue from the ShufflePlay entry's uri that can be used, instead of the name?

Comment (hufman) [reply app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:535]: In my test just now, the Shuffle Play's uri is the same as the `queueUri`, which might hold true in general and be a good way to drop that entry.

File: app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt (hunk 1 above)
Comment (Ben2356) [reply app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt]: Ah good point, removed this MusicAction

File: app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/BrowsePageView.kt (hunk 6 above)
Comment (Ben2356) [reply app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/BrowsePageView.kt]: Added a new "View Full Results" entry to the suggestion model along with a "..." entry at the end of the truncated suggestion list (due to the suggestionsModel only accurately showing the first 32 items) if the list is indeed truncated. Both of these new entries when clicked will open the SearchResultsView

File: app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/BrowsePageView.kt (hunk 7 above)
Comment (Ben2356) [reply app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/BrowsePageView.kt]: Yeah that's a good idea, updated this so now when the search results come back empty and Spotify is not being used the PlayFromSearch method will be called with the query input

**Comment by Ben2356** [comment]
> 
> 
> Can you also change the MusicPermissionsFragment layout to show a different string like `Ability to search Spotify` to emphasize the user benefit of clicking the button?

Updated the MusicPermissionsFragment string for the Spotify Web API authorization to be the suggested "Ability to search Spotify". It now fits in so much better with regards to the other permission text!

File: app/src/main/java/me/hufman/androidautoidrive/music/spotify/SpotifyAuthStateManager.kt (hunk 3 above)
Comment (Ben2356) [reply app/src/main/java/me/hufman/androidautoidrive/music/spotify/SpotifyAuthStateManager.kt:102]: Sadly this doesn't, this was actually done for a couple of reasons, mainly the logic for updating the token response is more fitting to the SpotifyAuthStateManager, the SpotifyWebApi class shouldn't be concerning itself with building out the proper token response to update the AuthState with. There's also a long story with breaking Powermock 😄 so I was trying to pull logic out of the SpotifyWebApi class to see if I could avoid using Powermock on it so this is the first step

File: app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt (hunk 2 above)
Comment (Ben2356) [reply app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt:533]: I added this refactoring since when an item such as an album is being browsed through the BrowseView the Shuffle Play button was present and needed to be removed. This problem was showing up when browsing a search result album type too

File: app/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -197,7 +197,7 @@
     <string name="txt_spotify_api_authorization_success">Spotify Web API authorized!</string>
     <string name="txt_spotify_api_authorization_canceled">Spotify Web API authorization canceled.</string>
     <string name="txt_spotify_api_authorization_failed">Spotify Web API authorization failed. Please try again.</string>
-    <string name="lbl_spotify_api_authorized">Spotify Web API Authorized</string>
+    <string name="lbl_spotify_api_authorized">Ability to search Spotify</string>
```
Comment (hufman) [inline app/src/main/res/values/strings.xml:213]: Ideally this should be a different string key, so that the (currently only 3) translations know to translate a fresh new string.

File: app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/SearchInputView.kt [hunk 11]
```diff
@@ -97,7 +121,11 @@ class SearchInputView(val state: RHMIState,
 			}
 
 			override fun onOk() {
-				browseController.showSearchResults(deferredSearchResults, inputComponent.getResultAction()?.asHMIAction())
+				if(!isSpotify && deferredSearchResults.isCompleted && deferredSearchResults.getCompleted()?.isEmpty() == true) {
+					browseController.playFromSearch(inputComponent.getResultAction()?.asHMIAction(), this.input)
+				} else {
```
Comment (hufman) [inline app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/SearchInputView.kt:126]: I like this good start! I think the `showSearchResults` shouldn't be the final `else` case, because relatively few music controllers will provide search results (though the first branch should catch unsearchable apps with a quickly-resolved empty list). I think `showSearchResults` should only be called when `!deferredSearchResults.isCompleted || deferredSearchResults.getCompleted()?.isNotEmpty` (to be used if it is actually loading or has already loaded some results), and the else branch should just not do anything, except perhaps raise an RHMIActionAbort to ignore any previous HMIAction target state.

**Comment by Ben2356** [comment]
> 
> 
> The SpotifyAppController.Connector has logic to trigger MusicAppDiscovery again after logging in with the control API to add the Browseable flag. Can you add similar logic to add the Searchable flag after the Web API is authenticated?

Yeah definitely, I updated the searchable flag after the SpotifyWebApi is authenticated along with the fixes to the MusicAppDiscovery class

File: app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt (hunk 9 above)
Comment (Ben2356) [reply app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt]: Ah good idea, I didn't realize the queueUri was the same as the shuffle play's uri! I removed the title matching and string and now only check if the mediaId == queueUri which is much better than before so now it is language agnostic

File: app/src/main/res/values/strings.xml (hunk 10 above)
Comment (Ben2356) [reply app/src/main/res/values/strings.xml]: Oh yeah you're right, reverted this change and added a new string with the "Ability to search Spotify" message that is being used

File: app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/SearchInputView.kt (hunk 11 above)
Comment (Ben2356) [reply app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/SearchInputView.kt]: Ah got it, changed that else to an else if with the conditions that it is not completed or if it is then it is not empty then showSearchResults will be called and omitted the else branch. I don't think we'll need to throw a RHMIActionAbort since the target state will not be set at all as it is only set through the playFromSearch() or showSearchResults method so it should remain unchanged in this else branch

Comment (hufman) [reply app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/SearchInputView.kt]: Excellent! It should be a rare situation when it would fall through to the else branch anyways.
Keep in mind that the HMIEvent's Target State isn't cleared during the app's lifetime and will remember whichever last state was last set. While it will start out as not going anywhere, any previous playFromSearch or showSearchResults will set this state and then future invocations of this branch won't remain on this InputState as intended. I think this case will be very rare, though, because the first two branches should cover most situations.

**Comment by hufman** [comment]
Looks great, thank you so much!
